<a href="https://colab.research.google.com/github/n8mcdunna/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/module1/Nathan_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [15]:
import pandas as pd
import numpy as np

url = 'https://archive.ics.uci.edu/ml/datasets/Automobile'
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data'
df = pd.read_csv(url)
column_names = ['symboling', 'normalized-losses', 'make', 'fuel-type', 
                'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 
                'engine-location', 'wheel-base', 'length', 'width', 'height', 
                'curb-weight', 'engine-type', 'num-of-cylinders', 'engine-size', 
                'fuel-system', 'bore', 'stroke', 'compression-ratio', 
                'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']
df.columns = column_names
df.head(20)

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0           3                 ?  alfa-romero  ...       21          27  16500
1           1                 ?  alfa-romero  ...       19          26  16500
2           2               164         audi  ...       24          30  13950
3           2               164         audi  ...       18          22  17450
4           2                 ?         audi  ...       19          25  15250
5           1               158         audi  ...       19          25  17710
6           1                 ?         audi  ...       19          25  18920
7           1               158         audi  ...       17          20  23875
8           0                 ?         audi  ...       16          22      ?
9           2               192          bmw  ...       23          29  16430
10          0               192          bmw  ...       23          29  16925
11          0               188          bmw  ...       21          28  20970
12          0               188          bmw  ...       21          28  21105
13          1                 ?          bmw  ...       20          25  24565
14          0                 ?          bmw  ...       16          22  30760
15          0                 ?          bmw  ...       16          22  41315
16          0                 ?          bmw  ...       15          20  36880
17          2               121    chevrolet  ...       47          53   5151
18          1                98    chevrolet  ...       38          43   6295
19          0                81    chevrolet  ...       38          43   6575

[20 rows x 26 columns]

In [16]:
df = df.replace('?', np.nan)
df.head(20)

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0           3               NaN  alfa-romero  ...       21          27  16500
1           1               NaN  alfa-romero  ...       19          26  16500
2           2               164         audi  ...       24          30  13950
3           2               164         audi  ...       18          22  17450
4           2               NaN         audi  ...       19          25  15250
5           1               158         audi  ...       19          25  17710
6           1               NaN         audi  ...       19          25  18920
7           1               158         audi  ...       17          20  23875
8           0               NaN         audi  ...       16          22    NaN
9           2               192          bmw  ...       23          29  16430
10          0               192          bmw  ...       23          29  16925
11          0               188          bmw  ...       21          28  20970
12          0               188          bmw  ...       21          28  21105
13          1               NaN          bmw  ...       20          25  24565
14          0               NaN          bmw  ...       16          22  30760
15          0               NaN          bmw  ...       16          22  41315
16          0               NaN          bmw  ...       15          20  36880
17          2               121    chevrolet  ...       47          53   5151
18          1                98    chevrolet  ...       38          43   6295
19          0                81    chevrolet  ...       38          43   6575

[20 rows x 26 columns]

In [17]:
df.isnull().sum()

symboling             0
normalized-losses    40
make                  0
fuel-type             0
aspiration            0
num-of-doors          2
body-style            0
drive-wheels          0
engine-location       0
wheel-base            0
length                0
width                 0
height                0
curb-weight           0
engine-type           0
num-of-cylinders      0
engine-size           0
fuel-system           0
bore                  4
stroke                4
compression-ratio     0
horsepower            2
peak-rpm              2
city-mpg              0
highway-mpg           0
price                 4
dtype: int64

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and use a 1-sample t-test to compare your estaimated sample mean to the population mean. Can you say that they are different? 

In [24]:
from scipy import stats

In [29]:
used_lot = df.sample(n= 10, random_state= 30)
# mean value for all of highway-mpg values
print('population ', df['highway-mpg'].mean())
# mean value for used car lot mpg
stats.ttest_1samp(a= used_lot['highway-mpg'], popmean= 30.769)

population  30.769607843137255


Ttest_1sampResult(statistic=-0.3411473017355997, pvalue=0.7408218810784865)

###The sample and the population are different since the p-valus is low at 0.74

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

In [35]:
print(stats.ttest_1samp(used_lot['highway-mpg'], 35))
used_lot['highway-mpg'].mean()


Ttest_1sampResult(statistic=-3.4187469470305465, pvalue=0.007643069993182772)


30.3

### T-test shows that the salesman's claim is false because of a very low p-value

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, take a sample of 100 cars. Run a t-test to verify the salesman's claim with your new larger sample. Do you reach the same conclusion as you did with the sample size of 10? 

In [36]:
used_lot_100_cars = df.sample(n= 100, random_state= 30)

In [37]:
stats.ttest_1samp(a= used_lot_100_cars['highway-mpg'], popmean= 35)

Ttest_1sampResult(statistic=-6.4827186446460106, pvalue=3.5481090083532426e-09)

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

###P-value for the sample of one hundred cars is much smaller than the sample of ten cars which shows the saleman's claim is even more false. T-statistic has also increased from -3.418 to -6.482 which shows that the null hypothesis of 35 is even more incorrect.


## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [64]:
congress_url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data'
col_names = ['party', 'handicaped-infants', 'water-projects', 
         'budget', 'physician-fee-freeze', 'el-salvador-aid', 
         'religious-groups', 'anti-satelitte-ban', 'aid-to-contras', 'mx-missile', 'immigration', 
         'synfuels', 'education', 'right-to-sue', 'crime', 'duty-free',
         'south-africe']
df1 = pd.read_csv(congress_url, names= col_names)

df1.head()

party handicaped-infants water-projects  ... crime duty-free south-africe
0  republican                  n              y  ...     y         n            y
1  republican                  n              y  ...     y         n            ?
2    democrat                  ?              y  ...     y         n            n
3    democrat                  n              y  ...     n         n            y
4    democrat                  y              y  ...     y         y            y

[5 rows x 17 columns]

In [65]:
# Issue: education
# dataset consists of the entire population

#taking a sample of the population
sample = df1.sample(n= 20, random_state= 42)
sample.head()

party handicaped-infants water-projects  ... crime duty-free south-africe
280    democrat                  n              n  ...     y         n            y
78     democrat                  y              y  ...     y         n            y
113  republican                  n              ?  ...     y         n            n
253  republican                  n              y  ...     y         n            y
324  republican                  n              y  ...     y         n            n

[5 rows x 17 columns]

In [66]:
# Cateorical encoding
df1 = df1.replace({'y': 1, 'n': 0, '?': np.nan})
df1.head()

party  handicaped-infants  ...  duty-free  south-africe
0  republican                 0.0  ...        0.0           1.0
1  republican                 0.0  ...        0.0           NaN
2    democrat                 NaN  ...        0.0           0.0
3    democrat                 0.0  ...        0.0           1.0
4    democrat                 1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [67]:
print(df1.shape)
df1.describe()

(435, 17)


handicaped-infants  water-projects  ...   duty-free  south-africe
count          423.000000      387.000000  ...  407.000000    331.000000
mean             0.442080        0.503876  ...    0.427518      0.812689
std              0.497222        0.500632  ...    0.495327      0.390752
min              0.000000        0.000000  ...    0.000000      0.000000
25%              0.000000        0.000000  ...    0.000000      1.000000
50%              0.000000        1.000000  ...    0.000000      1.000000
75%              1.000000        1.000000  ...    1.000000      1.000000
max              1.000000        1.000000  ...    1.000000      1.000000

[8 rows x 16 columns]

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

In [68]:
rep = df1[df1['party'] == 'republican']
dem = df1[df1['party'] == 'democrat']
dem.head()

party  handicaped-infants  water-projects  ...  crime  duty-free  south-africe
2  democrat                 NaN             1.0  ...    1.0        0.0           0.0
3  democrat                 0.0             1.0  ...    0.0        0.0           1.0
4  democrat                 1.0             1.0  ...    1.0        1.0           1.0
5  democrat                 0.0             1.0  ...    1.0        1.0           1.0
6  democrat                 0.0             1.0  ...    1.0        1.0           1.0

[5 rows x 17 columns]

In [62]:
# null hypothesis: congressional levels of support are equal to 0.5 (50%)
# alternative hypothesis: congressional levels of support are not equal to 0.5
# confidence level is 90%
stats.ttest_ind(dem['immigration'], rep['immigration'], nan_policy= 'omit')

Ttest_indResult(statistic=-1.7359117329695164, pvalue=0.08330248490425066)

In [ ]:
# Conclusion: A t-statistic of -1.73 and a p-value of 0.08 fails to reject the null hypothesis. 

# p-value < 0.01 reject
# p-value >= 0.01 fail to reject

Since the t-statistic is negative the republicans sample mean is larger which means republicans support immigration more than democrats.

In [88]:
# List comprehension for each issue comparing democrats to republicans
[(column, stats.ttest_ind(dem[column], rep[column], nan_policy= 'omit')) for column in df1.columns if column != 'party']

[('handicaped-infants',
  Ttest_indResult(statistic=9.205264294809222, pvalue=1.613440327937243e-18)),
 ('water-projects',
  Ttest_indResult(statistic=-0.08896538137868286, pvalue=0.9291556823993485)),
 ('budget',
  Ttest_indResult(statistic=23.21277691701378, pvalue=2.0703402795404463e-77)),
 ('physician-fee-freeze',
  Ttest_indResult(statistic=-49.36708157301406, pvalue=1.994262314074344e-177)),
 ('el-salvador-aid',
  Ttest_indResult(statistic=-21.13669261173219, pvalue=5.600520111729011e-68)),
 ('religious-groups',
  Ttest_indResult(statistic=-9.737575825219457, pvalue=2.3936722520597287e-20)),
 ('anti-satelitte-ban',
  Ttest_indResult(statistic=12.526187929077842, pvalue=8.521033017443867e-31)),
 ('aid-to-contras',
  Ttest_indResult(statistic=18.052093200819733, pvalue=2.82471841372357e-54)),
 ('mx-missile',
  Ttest_indResult(statistic=16.437503268542994, pvalue=5.03079265310811e-47)),
 ('immigration',
  Ttest_indResult(statistic=-1.7359117329695164, pvalue=0.08330248490425066)),
 

# Conclusions
### Democrats supported the handicapped issues more than republicans since the t-statistic is positive and the p-value is less than 0.01 at 1.61e-18.

### Republicans supported the physicisan fee freeze since the t-statistic was a large and negative value of -49.36 and the p-value was 1.99e-177 which is much less than the confidence level of 0.01.


### Democrats and republicans do not have much difference when the issue is water projects since the p-value for the 2-sample t-test is 0.929.



## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [ ]:
### YOUR WORK HERE

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [ ]:
### YOUR WORK HERE